src/charm.py

```diff
@@ -0,0 +1,477 @@
+
+        # Provide grafana dashboards over a relation interface
+        self._grafana_dashboards = GrafanaDashboardProvider(
+            self, relation_name="grafana-dashboard"
+        )
+        # TODO: in src/grafana_dashboards put an appropriate .tmpl file.
+
+    def _on_ingress_ready(self, event: IngressPerAppReadyEvent):
+        # TODO: Do I actually need to do anything here?
+        logger.info("This app's ingress URL: %s", event.url)
+
+    def _on_promtail_error(self, event):
+        logger.error(event.message)
+        # TODO: This is what the interface docs have, but it won't work with
+        # collect status!
+        self.unit.status = ops.BlockedStatus(event.message)
+
+    @property
+    def peers(self):
+        """Fetch the peer relation."""
+        return self.model.get_relation("takahe-peer")
+
+    def _on_collect_status(self, event: ops.CollectStatusEvent):  # noqa: C901
+        # Default to everything is ok.
+        event.add_status(ops.ActiveStatus())
+
+        # Is the config ok?
+        if not self.config["media-uri"].startswith(("local://", "s3://", "gcs://")):
+            event.add_status(ops.BlockedStatus("Invalid 'media-uri' config value."))
+        if not self.config.get("main-domain"):
+            event.add_status(ops.BlockedStatus("Please set the main domain with `juju config`"))
+        elif self.config["main-domain"] == "example.com":
+            event.add_status(ops.BlockedStatus("example.com is not a valid main domain"))
+
+        # Do we have the required storage?
+        if self.model.storages:
+            if "media-cache" not in self.model.storages:
+                event.add_status(ops.WaitingStatus("Waiting for media-cache"))
+            if self.config["media-uri"] == "local://" and "local-media" not in self.model.storages:
+                event.add_status(ops.WaitingStatus("Waiting for local-media"))
+        else:
+            event.add_status(ops.WaitingStatus("Waiting for storage"))
+
+        # Is the peer relation configured?
+        if self.peers:
+            # Do we have the required secrets?
+            try:
+                secret_id = self.peers.data[self.app]["secret-id"]
+            except KeyError:
+                event.add_status(ops.BlockedStatus("Waiting for peer secrets"))
+            else:
+                try:
+                    self.model.get_secret(id=secret_id).get_content()
+                except ops.SecretNotFoundError:
+                    event.add_status(ops.BlockedStatus("Waiting for secret key"))
+        else:
+            event.add_status(ops.WaitingStatus("Waiting for peer relation"))
+
+        # Check if we have everything we need to connect to the database.
+        if not self.dsn:
+            event.add_status(ops.BlockedStatus("Waiting for database relation"))
+
+        # Is ingress configured?
+        if not self.ingress.is_ready():
+            event.add_status(ops.WaitingStatus("Waiting for ingress."))
```

Comment:
This might be better as a Matrix or voice chat ...

For everything that works as a point-in-time check, I like it yes, and for the "nothing has happened but Juju would like a status update" case, I think it works well.

Where I'm less sure is the same place I've seen other charmers question things: what to do if something goes wrong in handling some other event. I feel like what I would really like is for `add_status` to be available on all hook events so that when something goes wrong I could just add in the appropriate blocked/waiting status, exit the handler, and then have everything work out.

I think you suggested somewhere else here that I could store information about the problem on the charm object and then extract it out in the collect-status handler. Is that a pattern that we're ok with? I know that at the moment the Charm object persists, but are we ok with charms relying on that, or would that be considered an anti-pattern?

I think, from what I've seen in other people's chats, that the other issue is when it's a problem that isn't necessarily going to be resolved by the time the next event's collect-status runs (and presumably can't be detected by collect-status, since otherwise I could always just ignore setting the status and exit). There was the recent question about storing status objects in stored state, for example.

I feel like collect-status provides *almost* everything that's needed, but not quite. But I'm not totally certain about that.

Reply:
Yeah, good discussion. Let's continue this discussion and maybe talk about some of the other points in our 1:1 next week (I've put it on the agenda).